### PR TITLE
Use multiindex in retained_map to use both hash map as well as ordered index

### PR DESCRIPTION
**Pull request by kleunen**
The following PR allows for optimization in retained_map by using two indices: hash map for direct key acccess, ordered index for wildcard access.

### CODE REVIEW

**Comment by redboltz** [comment]
There are conflicts with master. Please rebase it.

File: test/retained_topic_map.hpp [hunk 1]
```diff
@@ -216,46 +243,42 @@ class retained_topic_map {
     }
 
     // Increase the number of topics for this path
-    void increate_topics(std::vector<map_type_iterator> const &path) {
-        for (auto i : path) {
-            ++(i->second.count);
+    void increase_topics(std::vector<map_type_const_iterator> const &path) {
+        auto &direct_index = map.template get<direct_index_tag>();
+
+        for(auto i: path) {
```
Comment (redboltz) [inline test/retained_topic_map.hpp]:
```suggestion
        for (auto i : path) {
```

File: test/retained_topic_map.hpp [hunk 2]
```diff
@@ -216,46 +243,42 @@ class retained_topic_map {
...
-        ++next_node_id;
+        root = map.insert(path_entry(root_node_id, "", root_node_id)).first;
     }
 
     // Insert a value at the specified subscription path
-
-    template <typename V>
-    size_t insert_or_update(MQTT_NS::string_view topic, V&& value) {
+    size_t insert_or_update(MQTT_NS::string_view topic, Value const &value) {
+        auto &direct_index = map.template get<direct_index_tag>();
```
Comment (redboltz) [inline test/retained_topic_map.hpp]:
```suggestion
        auto& direct_index = map.template get<direct_index_tag>();
```

File: test/retained_topic_map.hpp [hunk 3]
```diff
@@ -280,8 +303,9 @@ class retained_topic_map {
     // Dump debug information
     template<typename Output>
     void dump(Output &out) {
-        for (auto const& i : map) {
-            out << i.first.first << " " << i.first.second << " " << (i.second.value ? "init" : "-") << " " << i.second.count << std::endl;
+        auto const &direct_index = map.template get<direct_index_tag>();
```
Comment (redboltz) [inline test/retained_topic_map.hpp]:
```suggestion
        auto const& direct_index = map.template get<direct_index_tag>();
```

File: test/retained_topic_map.hpp [hunk 4]
```diff
@@ -280,8 +303,9 @@ class retained_topic_map {
...
+        for (auto const& i: direct_index) {
```
Comment (redboltz) [inline test/retained_topic_map.hpp]:
```suggestion
        for (auto const& i : direct_index) {
```

File: test/retained_topic_map.cpp [hunk 5]
```diff
@@ -10,6 +10,9 @@
 
 #include "retained_topic_map.hpp"
 
+#include <iostream>
```
Comment (redboltz) [inline test/retained_topic_map.cpp:13]: I think that it is not required.
Test code shouldn't use iostream.

File: test/retained_topic_map.hpp (hunk 4 above)
Comment (kleunen) [reply test/retained_topic_map.hpp]: Wait a moment, because i was already updating according to coding guidelines

**Comment by kleunen** [comment]
Wait for CI to finish before starting the review, because if there is an issue in CI, this needs to be fixed first :)

**Comment by kleunen** [comment]
Some issue again with the hash calculation on basic_string_view on linux. Not sure how to fix this

**Comment by kleunen** [comment]
@redboltz Can you have a look, i just can't get the hash_value to work under linux for the retained_map ..

**Comment by redboltz** [comment]
Unfortunately, the error on CI / linux (7) is not reproduced on my local linux environment.

https://github.com/redboltz/mqtt_cpp/blob/master/include/mqtt/string_view.hpp#L47
I think that 

```cpp
std::size_t hash_value(basic_string_view<charT, traits> s) {
```

should be

```cpp
std::size_t hash_value(boost::basic_string_view<charT, traits> s) {
```

I've tested it on https://github.com/redboltz/mqtt_cpp/actions/runs/315501349 but the error isn't fixed.

-----

I think that it is inefficient test CI on my environment. You can't cancel and restart build.
I recommend that setup github actions on your github.
And

https://github.com/redboltz/mqtt_cpp/blob/master/.github/workflows/gha.yml#L8

replace

```yaml
    branches:
    - master
```


```yaml
    branches:
    - '*'
```

you can test on the CI when you push (no pull request reqiired)

**Comment by kleunen** [comment]
or maybe:

> ```c++
> namespace boost {
> std::size_t hash_value(boost::basic_string_view<charT, traits> s) {
>
> }
> ```

**Comment by kleunen** [comment]
> If you use boost { } , remove boost from basic_string_view is good. It is the same as boost header. I guess that there is no essential difference but just in case.

I will, 

and I think:
```
#if BOOST_VERSION < 106900
```

inside: 
```
#if BOOST_VERSION >= 106100
```

won't work either. Why did this work before ?

**Comment by kleunen** [comment]
CI / linux(7) compiles now, but test takes a very long time. Is it stuck ? Or just a very slow machine ?

**Comment by redboltz** [comment]
> CI / linux(7) compiles now, but test takes a very long time. Is it stuck ? Or just a very slow machine ?

See https://github.com/redboltz/mqtt_cpp/pull/697#issuecomment-712156019

Again, I recommend do test on **your** github environment.
It affects not only your PR but also all other PRs on my enviromnent.

**Comment by redboltz** [comment]
I just cancelled all queued job on my github actions expect your newest one from WebUI. I guess that the CI process start working soon.

I think that now `hash_value()` is defined in `MQTT_NS`. It is bad. You might have already noticed.

In this case, wandbox is useful to test. It can switch several version of boost.
https://wandbox.org/permlink/vztE0ACiz2UaNxkf

**Comment by redboltz** [comment]
BTW, https://github.com/redboltz/mqtt_cpp/actions is CI history. I needed to cancel one by one. It's boring work ;)  
But I really appreciate your effort.

**Comment by kleunen** [comment]
> BTW, https://github.com/redboltz/mqtt_cpp/actions is CI history. I needed to cancel one by one. It's boring work ;)
> But I really appreciate your effort.

I understand, no problem. I will close PR request now so no jobs are started when I push to my branch now.

**Comment by redboltz** [comment]
Thanks.
Your branch `retained_map_multiindex` is mapped to #697, so if you push some commits to `retained_map_multiindex`, then CI provess on redboltz/mqtt_cpp is automatically enqueued.

In this case, I use the following process. I hope this will help.

1. Create a new branch `retained_map_multiindex2`based on `retained_map_multiindex`.
2. Update gha.yml as follows:
```yaml
    branches:
    - '*'
```
3. Push `retained_map_multiindex2`. It triggered CI process on your environment. Nothing happens on redboltz/mqtt_cpp.
4. Fix bugs and push. You have whole control e.g.) cancel, re-start on your CI. It's convenient.
5. When all bugs are fixed (CI should be success), then remove the commit at the step.2 (`git rebase -i `retained_map_multiindex`).
6. Reflect `retained_map_multiindex2` commits to `retained_map_multiindex`.
7. push `retained_map_multiindex`. redboltz/mqtt_cpp CI process would start but CI should be success.

**Pull request closed without merging**